### PR TITLE
refactor(event_source): centralizing helper functions for query, header and base64

### LINKED ISSUE #2424: Tech debt: Centralize DictWrapper functions for query, header, json_data, etc.
### Why is this needed?

API Gateway (REST, HTTP), ALB, Lambda Function URL, and VPC Lattice unfortunately have differences in the payload. This leads to having nearly the same functionality in Event Source Data Classes to work around these differences and limitations.

These will eventually lead to a drift in implementation, making it harder to debug.

Latest example: https://github.com/awslabs/aws-lambda-powertools-python/pull/2358/files#diff-209f210ae09d053ce73d291e9f7b3869015b7f592f186be386518f152aa497adR56

### Which area does this relate to?

Event Source Data Classes

### Suggestion

Within `aws_lambda_powertools/utilities/data_classes/common.py`, we could have standalone functions like [`get_header_value`](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/utilities/data_classes/common.py#L93), and create new ones to handle JSON and Base64 deserialization.

This means updating code in several places where we might have HTTP Headers, HTTP Query Strings, HTTP JSON payload, Base64 payload -- SNS, SQS, Kinesis, API Gateway, AppSync, etc.

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (Python) Tenets](https://awslabs.github.io/aws-lambda-powertools-python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/awslabs/aws-lambda-powertools-java/), [TypeScript](https://github.com/awslabs/aws-lambda-powertools-typescript/), and [.NET](https://github.com/awslabs/aws-lambda-powertools-dotnet/)

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #2424 

### Changes

This PR unifies all the helper functions in one file to avoid drifts and fix minor problems in event source data classes.

### User experience

There is no change in the user experience.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/data_classes/common.py [hunk 1]
```diff
@@ -166,8 +150,7 @@ def get_query_string_value(self, name: str, default_value: Optional[str] = None)
         str, optional
             Query string parameter value
         """
-        params = self.query_string_parameters
-        return default_value if params is None else params.get(name, default_value)
+        return get_query_string_value(self.query_string_parameters, name, default_value)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/common.py]: prefer kwargs to ease reading/refactoring signature later

File: aws_lambda_powertools/utilities/data_classes/kafka_event.py [hunk 2]
```diff
@@ -69,18 +72,9 @@ def decoded_headers(self) -> Dict[str, bytes]:
...
-            return self.decoded_headers.get(name, default_value)
-        name_lower = name.lower()
-
-        return next(
-            # Iterate over the dict and do a case-insensitive key comparison
-            (value for key, value in self.decoded_headers.items() if key.lower() == name_lower),
-            # Default value is returned if no matches was found
-            default_value,
-        )
+        return get_header_value(self.decoded_headers, name, default_value, case_sensitive)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/kafka_event.py]: prefer kwargs to ease reading/refactor signature

File: aws_lambda_powertools/utilities/data_classes/rabbit_mq_event.py [hunk 3]
```diff
@@ -83,13 +82,13 @@ def data(self) -> str:
     @property
     def decoded_data(self) -> str:
         """Decodes the data as a str"""
-        return base64.b64decode(self.data.encode()).decode()
+        return base64_decode(self.data)
 
     @property
     def json_data(self) -> Any:
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/rabbit_mq_event.py:88]: note to future (not this PR): we should have a class decorator to handle these ubiquitous JSON and Base64 properties

File: aws_lambda_powertools/utilities/data_classes/active_mq_event.py [hunk 4]
```diff
@@ -125,7 +124,7 @@ def event_source_arn(self) -> str:
     @property
     def messages(self) -> Iterator[ActiveMQMessage]:
         for record in self["messages"]:
-            yield ActiveMQMessage(record)
+            yield ActiveMQMessage(record, json_deserializer=self._json_deserializer)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/active_mq_event.py:127]: fantastic catch! easy to miss.

**Comment by heitorlessa** [comment]
LGTM after you refactor to use kwargs over args in for `get_header_value` and `get_query_string_value`.

File: aws_lambda_powertools/utilities/data_classes/common.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/common.py]: Make total sense, Heitor!
Fixed

File: aws_lambda_powertools/utilities/data_classes/kafka_event.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/kafka_event.py]: Make total sense, Heitor!
Fixed

File: aws_lambda_powertools/utilities/data_classes/active_mq_event.py (hunk 4 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/active_mq_event.py:127]: Woww :laughing:

**Comment by heitorlessa** [comment]
Good to go.... still don't have my access fixed :/

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**